docs: Add Azure Active Directory docs

paulproteus opened the pull request:
Also:

- Make this page very wide, so that the images can be seen clearly.

- Add commented-out docs for Windows Server Active Directory, whose
  text & screenshots I haven't edited yet.

paulproteus commented:
Requesting a +1/-1 from @kentonv.

I know you suggested editing the screenshots, but I personally find them very readable as-is without e.g. arrows.

I also made one text change that seemed important to me to make, and yet I wasn't sure I understood the original text, so I wanted to know if it is an acceptable text change! Specifically:


In the original text, you wrote this:

> - Click the checkbox to complete configuration.
> - Click: "Assign accounts"
> - Choose "All Users" from the drop-down.

But I didn't see any screenshot with "Assign accounts" or "All users" in them. So I removed that  text in the instructions. But I might be missing something. Can you clarify?

I'll wait on the above before I merge this.

paulproteus commented:
FWIW:

https://github.com/sandstorm-io/sandstorm/blob/44b59aa7fd1b9f0df3990400748f0b0c9d1352e0/docs/administering/active-directory.md

is a good way to preview these changes!

paulproteus commented:
Question for @kentonv:

- In the Windows Server SAML docs, I notice that the Entity ID in Sandstorm is `example.sandcats.io` not `https://example.sandcats.io`. Is that intentional? Does it matter?

- In the context of the "Add Transform Claim Rule Wizard", you wrote this bullet point:

> - Under "Attribute store", choose "Active Directory".

But I didn't see any information about that in the screenshots.

Also @kentonv , for Azure, do we use the "old portal"? Presumably that means we'll need to re-do this layout & screenshot work in the future? Should we switch to the "new portal"?

ocdtrekkie commented:
I'm going to spend some time reviewing this, probably tonight, from the Windows system admin perspective.

ocdtrekkie commented:
"ActiveDirectory" needs a space in the Overview section.

paulproteus commented:
Yippee! Thanks @ocdtrekkie !

FWIW, make sure to review the current version, rather than the version that I link to above that's specific to a single GitHub commit ID.

https://github.com/sandstorm-io/sandstorm/blob/docs-for-azure-activedirectory/docs/administering/active-directory.md is a good way to get that.

Note also that the screenshots will be a little less wide when they go live in the docs. I can put up a test version of the generated docs if you like, but hopefully it doesn't matter much.

ocdtrekkie commented:
Yeah, I'm looking at the latest commit. I should use the Review feature, so I don't spam you with each one.

ocdtrekkie commented on docs/administering/active-directory.md:
```diff
@@ -0,0 +1,304 @@
+![Screenshot of this step](https://alpha-evgl4wnivwih0k6mzxt3.sandstorm.io/azure-ad-config/17.png)
+
+- In the Azure AD config, check the box: "Confirm that you have configured single sign-on as
+  described above." Clik the "next" arrow and enter your email address for service alerts from
```
"Click"

paulproteus replied:
Thanks, will fix

kentonv commented:
> In the Windows Server SAML docs, I notice that the Entity ID in Sandstorm is example.sandcats.io not https://example.sandcats.io. Is that intentional? Does it matter?

Sandstorm defaults to the bare hostname as the Entity ID. For Windows Server, it doesn't matter because it picks up the correct entity ID via the XML configuration blob -- the most important thing is that people don't change it after they've already configured the AD FS side. So I intentionally made sure the form field had its default value here, to avoid making people think they had to change it.

Azure, on the other hand, seems to require that the entity ID is a full URL.

> In the context of the "Add Transform Claim Rule Wizard", you wrote this bullet point:
>
> > Under "Attribute store", choose "Active Directory".
>
> But I didn't see any information about that in the screenshots.

It's the drop-down right in the middle of the dialog:

![](https://camo.githubusercontent.com/331ffdefbfb89de8b4920cff22a8e08b4fcc9fed/68747470733a2f2f616c7068612d6576676c34776e6976776968306b366d7a7874332e73616e6473746f726d2e696f2f77696e646f77732d7365727665722d61642d636f6e6669672f31322e706e67)

> Also @kentonv, for Azure, do we use the "old portal"? Presumably that means we'll need to re-do this layout & screenshot work in the future? Should we switch to the "new portal"?

AFAICT the new portal does not yet support configuring an arbitrary SAML endpoint. I noted this in the bullet points: "If this option is not displayed: Make sure you are using the classic portal. This option does not yet appear in the new portal. This option may require upgrading to Azure AD Premium."

kentonv commented on docs/administering/active-directory.md:
```diff
@@ -0,0 +1,304 @@
+
+Sandstorm can rely on AD to authenticate users.
+
+Active Directory exists as a cloud product called Windows Azure as well as a software product called
```
"exists as a" -> "exists as part of a"

Otherwise I feel like this is saying that Windows Azure and Windows Server are versions of Active Directory, rather than bundles that include versions of Active Directory...

kentonv commented on docs/administering/active-directory.md:
```diff
@@ -0,0 +1,304 @@
+
+Note that you can also set up Sandstorm to integrate with Active Directory using LDAP bind to
+authenticate users. We recommend Active Directory Federation Services or Microsoft Azure AD Single
+Sign-On instead, which use SAML. SAML has the advantage that users never type passwords into
```
I think "use" is correct -- he mentioned two different products earlier in the sentence.

ocdtrekkie replied:
You are in fact correct here. :/ This is just hard to read and sounds wrong because of Microsoft's tendency of naming schemes for things that are four words or longer.

paulproteus replied:
Thanks for noticing that this is confusing. I'll see if I can do a minor rephrasing.

kentonv commented on docs/administering/active-directory.md:
```diff
@@ -0,0 +1,304 @@
+Note that you can also set up Sandstorm to integrate with Active Directory using LDAP bind to
+authenticate users. We recommend Active Directory Federation Services or Microsoft Azure AD Single
+Sign-On instead, which use SAML. SAML has the advantage that users never type passwords into
+Sandstorm. If you must use LDAP instead, see the general [Sandstorm for Work documentation about
```
I think the main reason people will want to use SAML is not because Sandstorm sees their passwords with LDAP but rather because SAML allows AD full control over the login form, which enables things like multi-factor auth, Windows credentials login, etc. So I think that should be mentioned here...

paulproteus replied:
+1, see soon-to-come revised text

kentonv commented on docs/administering/active-directory.md:
```diff
@@ -0,0 +1,304 @@
+    - Under "SAML cert for above provider", copy the contents of the certificate you extracted in part
+      2, above.
+
+    - Use the full server URL as the entity ID. For example, https://example.sandcats.io
```
As mentioned elsewhere, it's important that people *don't* change the Entity ID at this point, otherwise they'll break the config that has already been synced into AD FS. The default entity ID does not include "https://" and works just fine here.

paulproteus replied:
+1, see soon-to-come revised text

kentonv commented on docs/administering/active-directory.md:
```diff
@@ -0,0 +1,304 @@
+
+To connect to Active Directory via SAML, you must enable Active Directory Federation Services (AD
+FS). We will assume you have already done this. We also assume that the Sandstorm server can
+reach the Active Directory server, e.g. they are on the same VPN, or firewall rules permit them
```
Actually, Sandstorm does not need to reach the AD server. Only the client's browser needs to reach it. With SAML all communication between the SP and IdP is through the browser.

The only server<->server communication happens during setup, where AD FS fetches the XML config blob from Sandstorm. So it's actually Sandstorm that needs to be accessible from AD. (Though, this could be worked around by the admin downloading the XML from Sandstorm and then uploading it to AD separately.)

paulproteus replied:
Fascinating. Thanks! I'll adjust the text accordingly.

paulproteus commented:
@ocdtrekkie and @kentonv - this is now ready for re-review. I believe I have addressed all feedback.

I'm still not sure about this aspect. I wrote in a comment at the top:

> In the original text, you [@kentonv] wrote this:
>
> > - Click the checkbox to complete configuration.
> > - Click: "Assign accounts"
> > - Choose "All Users" from the drop-down.
>
> But I didn't see any screenshot with "Assign accounts" or "All users" in them. So I removed that text in the instructions. But I might be missing something. Can you clarify?

Other than that I'm happy to ship this.

kentonv commented:
> Click the checkbox to complete configuration.

This should perhaps have been "click the checkmark button" -- it's a circular button in the lower-right of the window which would normally be labeled "OK" or "Finish" but some designer decided giving it a checkmark would be clearer.

> Click: "Assign accounts"

Oh, sorry. I should have mentioned this. So after you click the check to complete that stage of configuration, you get sent back to this screen:

![](https://camo.githubusercontent.com/d761f7b8084f463e9e5a3cef2a6309ddbf2e00a1/68747470733a2f2f616c7068612d6576676c34776e6976776968306b366d7a7874332e73616e6473746f726d2e696f2f617a7572652d61642d636f6e6669672f31302e706e67)

I didn't take a new screenshot since it's the same screen as before. This is where you click "Assign accounts".

> Choose "All Users" from the drop-down.

On this screen:

![](https://camo.githubusercontent.com/6311f1235037009d5ad72ca2a5bf733d331ca65b/68747470733a2f2f616c7068612d6576676c34776e6976776968306b366d7a7874332e73616e6473746f726d2e696f2f617a7572652d61642d636f6e6669672f31392e706e67)

The drop-down that says "Groups" needs to be changed to "All Users" (as it is in the following shot). Unfortunately I couldn't get a screenshot of the drop-down being open since the drop-down is rendered as a separate window thus isn't included in the shot. (You might have noticed the weird shot where the drop-box seems to be activated but there's no menu below it... that was my attempt to take a screenshot but it apparently didn't work.)

ocdtrekkie commented:
I am happy with all of the changes since my last review.

paulproteus commented:
OK! I think I've addressed all relevant feedback and will now merge. Post-merge feedback is OK too.
